functions/Test-DbaCompression.ps1

```diff
@@ -0,0 +1,371 @@
+SELECT @sqlmajorver = CONVERT(int, (@@microsoftversion / 0x1000000) & 0xff);
+
+IF @sqlmajorver = 9
+BEGIN
+	SET @sqlcmd = N'SELECT @StartDateOUT = login_time, @UpTimeOUT = DATEDIFF(mi, login_time, GETDATE()) FROM master..sysprocesses WHERE spid = 1';
+END
+ELSE
+BEGIN
+	SET @sqlcmd = N'SELECT @StartDateOUT = sqlserver_start_time, @UpTimeOUT = DATEDIFF(mi,sqlserver_start_time,GETDATE()) FROM sys.dm_os_sys_info';
+END
+
+SET @params = N'@StartDateOUT DATETIME OUTPUT, @UpTimeOUT VARCHAR(12) OUTPUT';
+
+EXECUTE sp_executesql @sqlcmd, @params, @StartDateOUT=@StartDate OUTPUT, @UpTimeOUT=@UpTime OUTPUT;
+
+--SELECT @StartDate AS Collecting_Data_Since, CONVERT(VARCHAR(4),@UpTime/60/24) + 'd ' + CONVERT(VARCHAR(4),@UpTime/60%24) + 'h ' + CONVERT(VARCHAR(4),@UpTime%60) + 'm' AS Collecting_Data_For
+
+SELECT 
+DBName = DB_Name()
+,[Schema] 
+,[Table_Name] 
+,[Index_Name] 
+,[Partition] 
+,[Index_ID] 
+,[Index_Type] 
+,[Percent_Scan] 
+,[Percent_Update] 
+,[ROW_estimate_Pct_of_orig] 
+,[PAGE_estimate_Pct_of_orig]
+,[Compression_Type_Recommendation] 
+,size_curKB = [size_cur]
+,size_reqKB = [size_req]
+FROM ##tmpCompression;
+
+DROP TABLE ##tmpCompression
+DROP TABLE ##tmpEstimateRow
+DROP TABLE ##tmpEstimatePage;"
+	}
+	
+	process {
+		
+		foreach ($instance in $SqlInstance) {
+			try {
+				Write-Message -Level VeryVerbose -Message "Connecting to $instance" -Target $instance
+				$server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $SourceSqlCredential
+			}
+			catch {
+				Stop-Function -Message "Failed to process Instance $Instance" -ErrorRecord $_ -Target $instance -Continue
+			}
+			
+            $Server.ConnectionContext.StatementTimeout = 0
+
+			#If IncludeSystemDBs is true, include systemdbs
+			#look at all databases, online/offline/accessible/inaccessible and tell user if a db can't be queried.
+			try {
+				if ($Database) {
+					$dbs = $server.Databases | Where-Object Name -In $Database
+				}
+				elseif ($IncludeSystemDBs) {
+					$dbs = $server.Databases | Where-Object Status -eq 'Normal'
+				}
+				else {
+					$dbs = $server.Databases | Where-Object { $_.status -eq 'Normal' -and $_.IsSystemObject -eq 0 }
+				}
+				
+				if (Was-Bound "ExcludeDatabase") {
+					$dbs = $dbs | Where-Object Name -NotIn $ExcludeDatabase
+				}
+			}
+			catch {
+				Stop-Function -Message "Unable to gather databases for $instance" -ErrorRecord $_ -Continue
+			}
+			
+
+			foreach ($db in $dbs) {
+				try {
+					Write-Message -Level Verbose -Message "Querying $instance - $db"
+					If ($db.status -ne 'Normal' -or $db.IsAccessible -eq $false) 
+                        {
+						Write-Message -Level Warning -Message "$db is not accessible." -Target $db
+                         
+						continue
+					    }
+                    #Execute query against individual database and add to output
+                    foreach ($row in ($db.ExecuteWithResults($sql)).Tables.Rows) 
```

Comment:
I've changed it to `foreach ($row in ($server.Query($sql, $db.Name)))` prettier and streams instead of assigning.